### PR TITLE
feat(jupyter): send Jupyter messaging metadata with `Deno.jupyter.broadcast`

### PR DESCRIPTION
**Pull request by manzt**
Will rebase following merge of #20710 in to `main`.

Exposes [`metadata`](https://jupyter-client.readthedocs.io/en/latest/messaging.html#metadata) to the `Deno.jupyter.broadcast` API.

```js
await Deno.jupyter.broadcast(msgType, content, metadata);
```

The metadata is required for [`"comm_open"`](https://github.com/jupyter-widgets/ipywidgets/blob/main/packages/schema/messages.md#instantiating-a-widget-object-1) for with `jupyter.widget` target. With this PR I can successfully communicate with the [`anywidget`](https://github.com/manzt/anywidget) frontend, sending frontend ESM via comms with updates :)

### Example 
Make sure you have anywidget installed:

```sh
pip install jupyterlab anywidget
```

```javascript
import { display } from "https://deno.land/x/display/mod.ts";

let comm_id = crypto.randomUUID();

await Deno.jupyter.broadcast("comm_open", {
    "comm_id": comm_id,
    "target_name": "jupyter.widget",
    "data": { 
        "state": {
            "_model_module": "anywidget",
            "_model_name": "AnyModel",
            "_model_module_version": "0.6.5",
            "_view_module": "anywidget",
            "_view_name": "AnyView",
            "_view_module_version": "0.6.5",
            "_view_count": null,
        }
    }
}, {
    "version": "2.1.0"
});

// Create an anywidget module (export a "render" function) for the front end
// The `model` is synchronized with `update` messages sent from the comm_msg
let esm = `export ${async function render({ model, el }) {
  let { default: confetti } = await import ("https://esm.sh/canvas-confetti@1.6.0");
  model.on("change:value", () => confetti({ angle: model.get("value") }));
  confetti();
}.toString()}`;

// Send initial widget state
await Deno.jupyter.broadcast("comm_msg", {
    "comm_id": comm_id,
    "data": { "method": "update", "state": { "_esm": esm, "_anywidget_id": comm_id } }
});

display({
    "application/vnd.jupyter.widget-view+json": {
        "version_major": 2,
        "version_minor": 1,
        "model_id": comm_id,
    }
})
```

```javascript
await Deno.jupyter.broadcast("comm_msg", {
    "comm_id": comm_id,
    "data": { "method": "update", "state": { "value": 50 } },
});

await new Promise(resolve => setTimeout(resolve, 1000));

await Deno.jupyter.broadcast("comm_msg", {
    "comm_id": comm_id,
    "data": { "method": "update", "state": { "value": 100 } },
});

await new Promise(resolve => setTimeout(resolve, 1000));
```

This is sooo cool!!

https://github.com/denoland/deno/assets/24403730/cebeeb7a-3be5-4e95-bdd5-5ebab864b647

### CODE REVIEW

**Comment by manzt** [comment]
cc: @rgbkrk

**Comment by manzt** [comment]
I'm sure others are more familiar with the latest for how to serialize functions / modules to run clientside. Currently I have this hack:

```javascript
// Create an anywidget module (export a "render" function) for the front end
// The `model` is synchronized with `update` messages sent from the comm_msg
let esm = `export ${async function render({ model, el }) {
  let { default: confetti } = await import ("https://esm.sh/canvas-confetti@1.6.0");
  model.on("change:value", () => confetti({ angle: model.get("value") }));
  confetti();
}.toString()}`;
```

But anywidget, just expects you to ship ESM that exports a function called `render`. The render function gives you access to  a `model` which you can listen for and emit updates to state for objects that live in the kernel.

**Comment by manzt** [comment]
https://github.com/denoland/deno/assets/24403730/c820fc93-05f7-4d80-9e02-be99eff566bb

File: cli/js/40_jupyter.js [hunk 1]
```diff
@@ -3,12 +3,14 @@
...
 function enableJupyter() {
-  denoNsUnstable.jupyter = {
-    async broadcast(msgType, content) {
-      await core.opAsync("op_jupyter_broadcast", msgType, content);
+  const {
+    op_jupyter_broadcast,
+  } = core.ensureFastOps();
+
+  globalThis.Deno.jupyter = {
+    async broadcast(msgType, content, metadata = {}) {
```
Comment (rgbkrk) [inline cli/js/40_jupyter.js:12]: 👍  for optional metadata

**Comment by rgbkrk** [comment]
This is awesome, I'm glad you were able to make this fly so quickly.

**Comment by rgbkrk** [comment]
> Would be awesome to be able to listen for messages from the frontend as well (i.e. enabling 2-way data binding). Not sure that is implemented yet.

That's #20592. It would be nice to have the Javascript API for it that we want to expose that is as close to one of the many web standard options (to be Deno-y).

You're making me realize we probably need to support [`buffers`](https://jupyter-client.readthedocs.io/en/latest/messaging.html#buffers) as the final argument of this big broadcast message.

**Comment by manzt** [comment]
>  It would be nice to have the Javascript API for it that we want to expose that is as close to one of the many web standard options (to be Deno-y).

Yes! I have some ideas.

> You're making me realize we probably need to support the [Jupyter message buffers](https://jupyter-client.readthedocs.io/en/latest/messaging.html#buffers) as the final optional argument of this big broadcast message.

Indeed.

The other option would be to have a single object for other possible message data.

```typescript
type DenoJupyter = typeof Deno & {
  jupyter: {
    broadcast(
      msg_type: string,
      content: { [key: string]: object },
      message?: {
        metadata?: Record<string, object>,
        buffers?: ArrayBuffer[]
      }
    ): Promise<void>;
  };
};
```

**Comment by bartlomieju** [comment]
I just merged https://github.com/denoland/deno/pull/20710, can you please rebase the PR @manzt?

**Comment by rgbkrk** [comment]
Famous last words and all, the jupyter spec will not extend the body of the message beyond content, metadata, and buffers.

I'm totally cool with `extra` though. 😄

**Comment by manzt** [comment]
> I just merged https://github.com/denoland/deno/pull/20710, can you please rebase the PR @manzt?

Done, @bartlomieju! TIL `git cherry-pick`.

**Review by bartlomieju — CHANGES_REQUESTED** [review]
Looks great, could you please update `cli/tsc/dts/lib.deno.unstable.d.ts` to match the new JS API?

**Comment by manzt** [comment]
done!

**Comment by manzt** [comment]
Implemented a `mod.ts` in anywidget @rgbkrk . (Think I might have messed up the denoland webhooks, since it's a monorepo and I'm not sure how to specify release for the `packages/anywidget/mod.ts` path).

```ts
import { widget } from "https://raw.githubusercontent.com/manzt/anywidget/main/packages/anywidget/mod.ts";

let model = await widget({
    // shared model between frontend and backend
    state: { value: 0 },
    // front-end render function
    render: ({ model, el }) => {
        let h1 = Object.assign(document.createElement("h1"), {
            innerText: "Value is " + model.get("value"),
        });
        model.on("change:value", () => {
            h1.innerText = "Value is " + model.get("value");
        });
        el.appendChild(h1);
    }
});

model
```

```ts
for (let i = 1; i <= 10; i++) {
    model.set("value", i);
    await new Promise(resolve => setTimeout(resolve, 500));
}
```

**Review by bartlomieju — APPROVED** [review]
LGTM, thank you @manzt, great improvement and I can't wait to see what you'll be able to do with it!

**Comment by rgbkrk** [comment]
Wrote a way to do imports with the anywidget in https://github.com/manzt/anywidget/pull/311

**Pull request merged**